scrapy/extensions/feedexport.py

```diff
@@ -186,86 +187,112 @@ def _store_in_thread(self, file):
-
     @classmethod
     def from_crawler(cls, crawler):
-        o = cls(crawler.settings)
-        o.crawler = crawler
+        o = cls(crawler)
         crawler.signals.connect(o.open_spider, signals.spider_opened)
         crawler.signals.connect(o.close_spider, signals.spider_closed)
         crawler.signals.connect(o.item_scraped, signals.item_scraped)
         return o
 
+    def __init__(self, crawler):
+        self.crawler = crawler
+        self.settings = crawler.settings
+        self.feeds = {}
+        self.slots = []
+
+        if not self.settings['FEEDS'] and not self.settings['FEED_URI']:
+            raise NotConfigured
+
+        # Begin: Backward compatibility for FEED_URI and FEED_FORMAT settings
+        if self.settings['FEED_URI']:
+            warnings.warn(
+                'The `FEED_URI` and `FEED_FORMAT` settings have been deprecated in favor of '
+                'the `FEEDS` setting. Please see the `FEEDS` setting docs for more details',
+                category=ScrapyDeprecationWarning, stacklevel=2,
+            )
+            self.feeds[self.settings['FEED_URI']] = feed_complete_default_values_from_settings(
+                {'FORMAT': self.settings.get('FEED_FORMAT', 'jsonlines')},
+                self.settings
+            )
+        # End: Backward compatibility for FEED_URI and FEED_FORMAT settings
+
+        for uri, feed in self.settings.getdict('FEEDS').items():
+            self.feeds[uri] = feed_complete_default_values_from_settings(
+                {key.strip().upper(): value for key, value in feed.items()},
+                self.settings
+            )
+
+        self.storages = self._load_components('FEED_STORAGES')
+        self.exporters = self._load_components('FEED_EXPORTERS')
+        for uri, feed in self.feeds.items():
+            if not self._storage_supported(uri):
+                raise NotConfigured
+            if not self._exporter_supported(feed['FORMAT']):
+                raise NotConfigured
+
     def open_spider(self, spider):
-        uri = self.urifmt % self._get_uri_params(spider)
-        storage = self._get_storage(uri)
-        file = storage.open(spider)
-        exporter = self._get_exporter(file, fields_to_export=self.export_fields,
-            encoding=self.export_encoding, indent=self.indent)
-        if self.store_empty:
-            exporter.start_exporting()
-            self._exporting = True
-        self.slot = SpiderSlot(file, exporter, storage, uri)
+        for uri, feed in self.feeds.items():
+            uri = uri % self._get_uri_params(spider, feed['URI_PARAMS'])
+            storage = self._get_storage(uri)
+            file = storage.open(spider)
+            exporter = self._get_exporter(
+                file=file,
+                feed_format=feed['FORMAT'],
+                fields_to_export=feed['FIELDS'],
+                encoding=feed['ENCODING'],
+                indent=feed['INDENT'],
```

Comment:
I was writing an answer for a [StackOverflow question](https://stackoverflow.com/q/57996812/939364) about customizing `XmlItemExporter` parameters, and I think maybe we should pave the way to make this code easier to extend in the future.

For example, if we use lowercase settings in `FEEDS`, and we deprecate `feed_format` and `fields_to_export` in item exporters and replace them with `format` and `fields`, we can use `exporter = self._get_exporter(file=file, **feed)` here.

That in turn allows users to take advantage of additional item exporter `__init__` parameters that are currently not exposed as settings, such as XmlItemExporter’s `item_element` and `root_element`, which would make it possible to solve the use case described in the [StackOverflow question](https://stackoverflow.com/q/57996812/939364) using only the `FEEDS` setting.

Then we would just need to document those additional `__init__` parameters that each item exporter supports.

Reply:
Great idea, thanks :smile:

PR updated, just a few comments:
* The `feed_format` name was introduced earlier in this PR, to avoid overriding the [`format` built-in function](https://docs.python.org/3/library/functions.html#format). It was only used internally in the `extensions/feedexport.py` file. I reverted that change, `format` shouldn't be a problem since the binding applies only to "private" (i.e. underscored) classes and methods.
* I'm hesitant about which name to use to configure the fields in the `FEEDS` setting. On the one hand, `fields` seems prettier and simpler than `fields_to_export`, on the other hand using `fields_to_export` would be consistent with the existing exporter init parameter and would allow us to make the follow-up change in a backward compatible manner. Thoughts?

Reply:
If we want to eventually rename the internal parameter to `fields`, I think it’s best we name the user-facing parameter `fields` already.

That way, once we change the internal field, users don’t need to change their `FEEDS` setting. Only people using the private `_get_exporter` method will need to update their code.

Reply:
True, but `fields_to_export` is [already documented](https://docs.scrapy.org/en/1.7/topics/exporters.html?highlight=fields_to_export#baseitemexporter).
Since the `FEEDS` setting is new, we could adapt to the current name and use `FEEDS.fields_to_export` instead of `FEEDS.fields` and prevent a backward-compatible change in the future.

Reply:
Oh, so stick to `fields_to_export`, with no plans of eventually changing it to just `fields`.

Works for me :+1: